inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py

```diff
@@ -0,0 +1,363 @@
+sys.path.insert(0, '/home/runner/GITHUB_ACTION_RUNNERS/_work/genai_ag_intel-inb-manageability/genai_ag_intel-inb-manageability/inbm-lib/')
+sys.path.insert(0, '/home/runner/GITHUB_ACTION_RUNNERS/_work/genai_ag_intel-inb-manageability/genai_ag_intel-inb-manageability/inbm/dispatcher-agent/dispatcher/source/')
+import pytest
+from unittest.mock import patch, MagicMock
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationAddSourceParameters
+from unittest.mock import patch, mock_open
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationSourceList
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationRemoveSourceParameters, SourceError
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, SourceError, ApplicationUpdateSourceParameters
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceParameters, SourceError
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceError
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerAdd
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerAdd:
+    @pytest.fixture
+    def mock_dispatcher_broker(self):
+        return MagicMock()
+
+    @pytest.fixture
+    def mock_application_source_manager(self, mock_dispatcher_broker):
+        return UbuntuApplicationSourceManager(mock_dispatcher_broker)
+
+    @pytest.mark.parametrize('gpg_key_name, gpg_key_uri, source_list_file_name, sources', [
+        ('test_key', 'http://test.com/key', 'test_file', ['deb http://test.com/ubuntu bionic main']),
+        (None, None, 'test_file', ['deb http://test.com/ubuntu bionic main']),
+        ('test_key', 'http://test.com/key', 'test_file', []),
+    ])
+    @patch('dispatcher.source.ubuntu_source_manager.verify_source')
+    @patch('dispatcher.source.ubuntu_source_manager.add_gpg_key')
+    @patch('dispatcher.source.ubuntu_source_manager.create_file_with_contents')
+    def test_add(self, mock_create_file_with_contents, mock_add_gpg_key, mock_verify_source, mock_application_source_manager, gpg_key_name, gpg_key_uri, source_list_file_name, sources):
+        # Arrange
+        parameters = ApplicationAddSourceParameters(gpg_key_name=gpg_key_name, gpg_key_uri=gpg_key_uri, source_list_file_name=source_list_file_name, sources=sources)
+
+        # Act
+        mock_application_source_manager.add(parameters)
+
+        # Assert
+        if gpg_key_name and gpg_key_uri:
+            mock_verify_source.assert_called_once()
+            mock_add_gpg_key.assert_called_once_with(gpg_key_uri, gpg_key_name)
+        mock_create_file_with_contents.assert_called_once_with('/etc/apt/sources.list.d/' + source_list_file_name, sources)
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerList
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerList:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/pull/4#)

<h4>For feedback purposes, please include both emoticon-based and qualitative feedback in your reply to this comment.</h4>

To go back to Pull Request Summary table, Click on the [link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/pull/4#)